### PR TITLE
BUG: fix+test op(NaT, ndarray), also simplify

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/_libs/tslibs/nattype.pyx [hunk 1]
```diff
@@ -103,61 +106,102 @@ cdef class _NaT(datetime):
         if ndim == -1:
             return _nat_scalar_rules[op]
 
+        if util.is_array(other):
```
Comment (jreback) [inline pandas/_libs/tslibs/nattype.pyx]: can you make these if/elif/else

File: pandas/_libs/tslibs/nattype.pyx [hunk 2]
```diff
@@ -103,61 +106,102 @@ cdef class _NaT(datetime):
...
-            # to subclass; allows us to @final(_Timestamp.__add__)
-            return NotImplemented
-        return c_NaT
+
+        elif is_integer_object(other) or util.is_period_object(other):
+            # For Period compat
+            # TODO: the integer behavior is deprecated, remove it
+            return c_NaT
+
+        elif util.is_array(other):
```
Comment (jreback) [inline pandas/_libs/tslibs/nattype.pyx]: what if there is a tz?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/nattype.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/nattype.pyx]: this is specifically for ndarray, so there cant be a tz

File: pandas/_libs/tslibs/nattype.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/nattype.pyx]: sure

Comment (jbrockmendel) [reply pandas/_libs/tslibs/nattype.pyx]: changed+green

**Comment by jbrockmendel** [comment]
@jreback gentle ping; a few open PRs are blockers for test consolidation etc

**Comment by jreback** [comment]
will try over weekend

File: pandas/_libs/tslibs/nattype.pyx [hunk 3]
```diff
@@ -103,61 +106,102 @@ cdef class _NaT(datetime):
         if ndim == -1:
             return _nat_scalar_rules[op]
 
-        if ndim == 0:
+        elif util.is_array(other):
```
Comment (jreback) [inline pandas/_libs/tslibs/nattype.pyx]: the ```getattr(other, 'ndim', -1)``` is very non-standard (as we almost always make this return 0 by default). can you just use a hasattr check? as the ndim == 0 branch is also never hit as a 0-line scalar hits the is_array branch. (maybe make that clear in a comment). also can pull out line 115 and make it part of the elif

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/nattype.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/nattype.pyx]: yah its weird, but it predates me.  the logic here is kind of weird; id rather take the time to look at it more carefully in a separate pass

File: pandas/tests/scalar/test_nat.py [hunk 4]
```diff
@@ -435,6 +456,28 @@ def test_nat_arithmetic_td64_vector(op_name, box):
...
 
 
+@pytest.mark.parametrize(
+    "dtype,op,out_dtype",
+    [
+        ("datetime64[ns]", operator.add, "datetime64[ns]"),
+        ("datetime64[ns]", roperator.radd, "datetime64[ns]"),
+        ("datetime64[ns]", operator.sub, "timedelta64[ns]"),
+        ("datetime64[ns]", roperator.rsub, "timedelta64[ns]"),
+        ("timedelta64[ns]", operator.add, "datetime64[ns]"),
```
Comment (jreback) [inline pandas/tests/scalar/test_nat.py]: shouldn't these last for yield timedelta64[ns]?

Comment (jbrockmendel) [reply pandas/tests/scalar/test_nat.py]: No.

dt64, radd --> dt64 +  Nat --> only valid if NaT is treated as td64 --> dt64
dt64, sub --> NaT - dt64 --> only valid if NaT is treated as dt64 --> td64
dt64, rsub --> dt64 - NaT --> valid either way, default to treat as dt64 --> td64
td64, add --> NaT + td64 --> valid either way, defalt to treat as dt64 --> dt64

**Review by jreback — APPROVED** [review]

File: pandas/tests/scalar/test_nat.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/scalar/test_nat.py]: ok, yeah these rules are ok, but if you can put them down somewhere explicity in the code

**Pull request merged**